centralstore/redis_store.go

```diff
@@ -558,52 +508,43 @@ func (r *RedisBasicStore) RecordTraceDecision(ctx context.Context, trace *Centra
 	return nil
 }
 
-func (r *RedisBasicStore) getTraceState(ctx context.Context, conn redis.Conn, traceID string) (state CentralTraceState) {
-	ctx, span := r.Tracer.Start(ctx, "getTraceState")
+func (r *RedisBasicStore) getTraceStates(ctx context.Context, conn redis.Conn, states map[string]CentralTraceState) error {
+	ctx, span := r.Tracer.Start(ctx, "getTraceStates")
 	defer span.End()
-	defer otelutil.AddSpanField(span, "state", state)
 
-	otelutil.AddSpanField(span, "trace_id", traceID)
-	if tracerec, _, found := r.DecisionCache.Test(traceID); found {
-		// it was in the decision cache, so we can return the right thing
+	var cacheHitCount int
+	notFound := make([]string, 0, len(states))
+	for traceID := range states {
+		tracerec, _, found := r.DecisionCache.Test(traceID)
+		if !found {
+			// if the trace is not in the cache, we need to retrieve its state from redis
+			notFound = append(notFound, traceID)
+			continue
+		}
+
+		cacheHitCount++
 		if tracerec.Kept() {
-			state = DecisionKeep
+			states[traceID] = DecisionKeep
 		} else {
-			state = DecisionDrop
+			states[traceID] = DecisionDrop
 		}
-
-		otelutil.AddSpanFields(span, map[string]interface{}{
-			"in_cache": true,
-		})
-
-		return state
-	}
-
-	if r.states.exists(ctx, conn, Collecting, traceID) {
-		return Collecting
-	}
-
-	if r.states.exists(ctx, conn, DecisionDrop, traceID) {
-		return DecisionDrop
-	}
-
-	if r.states.exists(ctx, conn, DecisionKeep, traceID) {
-		return DecisionKeep
 	}
+	otelutil.AddSpanField(span, "cache_hit_count", cacheHitCount)
 
-	if r.states.exists(ctx, conn, AwaitingDecision, traceID) {
-		return AwaitingDecision
+	if cacheHitCount == len(states) {
+		return nil
 	}
 
-	if r.states.exists(ctx, conn, ReadyToDecide, traceID) {
-		return ReadyToDecide
+	results, err := r.traces.getTraceStates(ctx, conn, notFound)
```

Comment:
We end up looping through all the traceids in `notFound` inside `r.traces.getTraceStates`.  Instead of building a second list and looping through it again, can we do the logic in `r.traces.getTraceStates` in the above for loop if the traceid was not found?

Reply:
Or, even better, is it possible to collect all the not found traceIDs like you are and then make a single request to redis for all the states of thoses traceids in 1 query?

Reply:
I pipelined them into one request